### ISSUE TITLE
LSTM / recurrent policy support

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/thrust#262]
## Summary
Add a recurrent (LSTM/GRU) policy option to the Burn policy layer. Roadmap Phase 3 item; currently only feedforward `MlpBurnPolicy` exists (no `recurrent`/`hidden_state` anywhere in `src/`).

## Motivation
Recurrent policies are needed for partially-observable environments (e.g. memory tasks, framestack-free Atari, sequential coordination). This is a prerequisite for several POMDP-style benchmarks.

## Scope / pointers
- Policy layer: `src/policy/` (alongside `MlpBurnPolicy`).
- Rollout storage must thread per-step hidden state: `src/buffer/rollout/`.
- On-policy trainers consume sequences with resets at episode boundaries: `src/train/ppo/`, `src/train/a2c/`.

## Acceptance criteria
- [ ] A recurrent policy type (LSTM, optionally GRU) implementing the existing policy trait.
- [ ] Hidden state correctly reset on `done` and threaded through rollouts.
- [ ] At least one example/test on a partially-observable task showing it learns.
- [ ] Seeded, reproducible init consistent with `seeded_init`.

## Notes
Design the hidden-state threading carefully — this touches the rollout buffer contract. Consider whether a short design note (or epic decomposition) is warranted before implementation.

---

## Design Note & Child Issues

Per the curator's complexity assessment, this is an **epic**: the deliverable is a design note plus a decomposition into scoped implementation issues. The design note — `docs/RECURRENT_POLICY_DESIGN.md` — decides the two load-bearing questions (buffer strategy: new `RecurrentRolloutBuffer`; minibatch strategy: full-sequence / Strategy A) plus the trainer API and POMDP env, grounded in the actual code and the Burn 0.21 LSTM API.

- [ ] #285 — Phase 1: `LstmBurnPolicy` + seeded LSTM init
- [ ] #286 — Phase 2: `RecurrentRolloutBuffer` + full-sequence sampler
- [ ] #287 — Phase 3: `RecurrentPPOTrainer` + MaskedCartPole POMDP + learning example

Deferred (reserved by the design, not filed): Strategy B fixed-length subsequences, GRU variant, T-Maze memory env, V-trace + recurrent integration.

**Comment by rjwalters** [comment]
## Curator Analysis: LSTM / Recurrent Policy Support

Curating this issue per user assignment. The original Notes section asks "consider whether a short design note (or epic decomposition) is warranted." After thorough codebase investigation, my answer is: **yes, decomposition is warranted.** Details below.

---

## Technical Context

### Policy Trait Surface

There is no formal `Policy` trait in this codebase. Each policy (`MlpBurnPolicy`, `SnakeCnnBurnPolicy`, `QNetworkBurn`, `SacActor`, etc.) is a concrete `#[derive(Module)]` struct. The PPO trainer does not dispatch through a trait; instead it accepts a closure:

```rust
// src/train/ppo/trainer.rs — PPOTrainerBurn::train_step
F: FnMut(&P, Tensor<B, 2>, Tensor<B, 1, Int>) -> (Tensor<B, 1>, Tensor<B, 1>, Tensor<B, 1>)
```

The bound is `P: AutodiffModule<B>`. This closure-based design means an `LstmBurnPolicy` does not need a new trait; it just needs the same Burn `Module` derive and compatible entry-point methods. However, the closure's current obs shape — `Tensor<B, 2>` i.e. `[batch, obs_dim]` — is **incompatible with recurrent evaluation**, which requires sequences. Changing it is a breaking change to the trainer API.

### Burn 0.21 LSTM / GRU Availability

Confirmed in `~/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/burn-nn-0.21.0/src/modules/rnn/`:
- `lstm.rs` — `Lstm<B>`, `LstmState<B, D>`, `LstmConfig`
- `gru.rs` — `Gru<B>`, `GruState<B, D>`, `GruConfig`

The LSTM forward signature:

```rust
pub fn forward(
    &self,
    batched_input: Tensor<B, 3>,         // [batch, seq_len, input_size] when batch_first=true
    state: Option<LstmState<B, 2>>,      // cell + hidden, each [batch, hidden_size]
) -> (Tensor<B, 3>, LstmState<B, 2>)    // full output sequence + final state
```

`LstmConfig` supports `batch_first` (default true), `initializer` (Xavier Normal default), optional cell clipping, and input-forget coupling. GRU is simpler (one state tensor, not two).

### Rollout Buffer: No Hidden State Storage

`src/buffer/rollout/storage.rs` — `RolloutBuffer` fields today:

```
observations: Vec<Vec<Vec<f32>>>   // [num_steps, num_envs, obs_dim]
actions:      Vec<Vec<i64>>        // [num_steps, num_envs]
rewards:      Vec<Vec<f32>>        // [num_steps, num_envs]
values:       Vec<Vec<f32>>        // [num_steps, num_envs]
log_probs:    Vec<Vec<f32>>        // [num_steps, num_envs]
terminated:   Vec<Vec<bool>>       // [num_steps, num_envs]
truncated:    Vec<Vec<bool>>       // [num_steps, num_envs]
advantages:   Vec<Vec<f32>>        // [num_steps, num_envs]
returns:      Vec<Vec<f32>>        // [num_steps, num_envs]
```

There is no hidden-state storage. Adding it requires either a new `RecurrentRolloutBuffer` struct or extending the existing one with optional hidden/cell fields — both have downstream effects on every buffer consumer.

### Minibatch Sampling: Incompatible with RNNs

`src/buffer/rollout/sampling.rs` — `generate_minibatch_indices` and the PPO trainer's `generate_minibatch_indices_with_rng` in `src/train/ppo/loss.rs` both **shuffle all `num_steps * num_envs` timestep indices** and chunk by `batch_size`. This destroys temporal ordering entirely.

Recurrent PPO requires one of:

- **Strategy A (full-sequence)**: Treat each environment's rollout as one sequence of length `T`. Minibatch = N envs × T steps as a `[N, T, obs_dim]` rank-3 tensor. Reset hidden state at `terminated` flags during the LSTM forward. No stored states needed during training.
- **Strategy B (fixed-length subsequences, à la SB3-RecurrentPPO)**: Divide each env trajectory into fixed-length chunks of length `seq_len`. Store the LSTM state at each chunk boundary. Sample K chunks per minibatch; initialize LSTM from stored boundary state. More flexible but requires storing states in the buffer.

The choice has significant implications for buffer design, trainer API, and whether hidden states need to be stored at all.

### V-trace Status

`src/buffer/rollout/vtrace.rs` exists on `origin/main` (confirmed via `git ls-tree`). It does not appear in the local working tree because the local checkout is behind remote. This module is compatible with recurrent policies in principle (the IMPALA paper uses LSTM policies), but that integration is a separate concern.

### Seeded Init for LSTM

`src/policy/seeded_init.rs` provides `seeded_orthogonal` and `seeded_kaiming_uniform` for `Vec<f32>` weight buffers. For the LSTM gates (4 weight matrices of shape `[d_hidden, d_input + d_hidden]`), the same seeded helper can be used to replace Burn's unseeded Xavier Normal. The gate indexing pattern from `mlp.rs::derive_layer_seed` applies directly.

### Partially-Observable Environments

No POMDP/memory task exists in `src/env/games/`. All current environments expose the full simulator state as observation. A new env is required.

Recommended options:
- **MaskedCartPole**: Drop `cart_velocity` and `pole_angular_velocity` from the 4D observation, leaving only `[cart_position, pole_angle]`. The environment is still fully simulated; the policy only lacks velocity information. An LSTM can integrate position over time to recover velocity. Simplest to implement (~150 LOC extending `src/env/games/cartpole.rs`).
- **T-Maze**: Classic memory proof-of-concept — agent sees a cue at the start of a corridor and must remember it at the decision point. More convincing demonstration but requires new geometry logic (~250 LOC new file).

MaskedCartPole is recommended for its simplicity; T-Maze for its scientific rigor.

---

## Complexity Assessment

**This exceeds the single-PR threshold. Decomposition is required.**

The changes span multiple non-trivial systems:

| Component | Files | Estimated LOC | Complexity |
|-----------|-------|--------------|------------|
| `LstmBurnPolicy` module | `src/policy/lstm.rs` (new) | ~300 | Medium — straightforward Burn Module |
| Rollout buffer extension | `src/buffer/rollout/storage.rs` | ~150 | High — contract-breaking change affecting all buffer consumers |
| Sequence-preserving sampler | `src/buffer/rollout/sampling.rs` | ~120 | High — fundamentally different algorithm |
| Recurrent PPO trainer | new `src/train/ppo/recurrent_trainer.rs` | ~350 | High — new trainer or signature change |
| POMDP environment | `src/env/games/masked_cartpole.rs` (new) | ~200 | Medium |
| End-to-end example/test | `examples/` | ~150 | Low |
| Seeded LSTM init | `src/policy/seeded_init.rs` | ~50 | Low |

Total: ~1300 LOC across ~8 files, with two contract-breaking changes (rollout buffer, trainer evaluate_fn).

The buffer contract change and the sequence-preserving sampler are load-bearing design decisions that cannot be finalized without answering the Strategy A vs. B question. These decisions also determine whether the LSTM state needs to be stored in the buffer at all. Getting this wrong means reworking both the buffer and the trainer.

**Recommendation: Follow the pattern of epics #265 and #266 — file a design note first, then decompose into 3–4 scoped implementation issues.**

The design note should answer:
1. **Buffer strategy**: New `RecurrentRolloutBuffer` (isolated, no breakage to existing buffer) vs. extending `RolloutBuffer` with optional hidden-state fields?
2. **Minibatch strategy**: Full-sequence (Strategy A) vs. fixed-length subsequences with stored boundary states (Strategy B)?
3. **Trainer API**: New `RecurrentPPOTrainer<B, P, O>` (preferred, avoids breaking current `evaluate_fn` closure signature) vs. parameterizing the existing trainer?
4. **POMDP env**: MaskedCartPole vs. T-Maze?

---

## Proposed Phase Breakdown (if approved as epic)

**Phase 1 — Design Note** (~2 hours)
Write a `docs/RECURRENT_POLICY_DESIGN.md` answering the four questions above, with a concrete API sketch for `RecurrentRolloutBuffer` and `RecurrentPPOTrainer`. No code changes.

**Phase 2 — `LstmBurnPolicy` + seeded init** (~4 hours)
Implement `LstmBurnPolicy<B>` in `src/policy/lstm.rs` with:
- LSTM trunk + policy head (logits) + value head
- `forward_step(obs, state) -> (logits, value, new_state)` for rollout collection
- `evaluate_sequences(obs_seq, actions, initial_states) -> (log_probs, entropy, values)` for training
- Seeded init via derived seeds on gate weight matrices

**Phase 3 — `RecurrentRolloutBuffer` + sequence sampler** (~4 hours)
- New `RecurrentRolloutBuffer` that stores `hidden` and `cell` state fields per step
- `add_hidden_state(step, env_id, h, c)` method
- New sequence-preserving sampler returning `[batch, seq_len, obs_dim]` tensors plus initial-state bundles
- All existing `RolloutBuffer` callers unaffected

**Phase 4 — `RecurrentPPOTrainer` + POMDP env + example** (~5 hours)
- New `RecurrentPPOTrainer<B, P, O>` that consumes sequences from `RecurrentRolloutBuffer`
- New partially-observable environment (`MaskedCartPole` or T-Maze)
- Example/integration test showing the recurrent policy learns on the POMDP task

---

## Acceptance Criteria

- [ ] `LstmBurnPolicy<B>` implements Burn's `Module` and `AutodiffModule` derives.
- [ ] A recurrent policy (LSTM, optionally GRU) is constructable via an equivalent of `MlpBurnConfig` with a `seed: Option<u64>` field; two constructions with the same seed produce bit-identical weights.
- [ ] Hidden state is correctly reset when `terminated = true` and threaded between steps when `terminated = false`.
- [ ] `RecurrentRolloutBuffer` (or extended `RolloutBuffer`) stores per-step `(h_t, c_t)` without breaking existing feedforward buffer consumers.
- [ ] Minibatch sampling respects temporal ordering — sequences are not split across episode boundaries; `terminated` flags trigger hidden-state resets during training.
- [ ] A new partially-observable environment exists (`MaskedCartPole` or T-Maze) in `src/env/games/`.
- [ ] At least one integration test or example demonstrates a recurrent policy achieving >195 mean return on `MaskedCartPole` (or equivalent performance criterion on the chosen POMDP) — the learning signal that feedforward policies cannot match on this task.
- [ ] No regressions on existing feedforward PPO/A2C tests.

---

## Affected Files

**To modify (existing on `origin/main`):**
- `src/policy/mod.rs` — add `pub mod lstm` under `#[cfg(feature = "training")]`
- `src/buffer/rollout.rs` — re-export new recurrent buffer type
- `src/buffer/rollout/storage.rs` — extend with optional hidden state fields, or keep untouched if using a separate `RecurrentRolloutBuffer`
- `src/buffer/rollout/sampling.rs` — add sequence-preserving sampler alongside existing random sampler
- `src/train/ppo/trainer.rs` — optionally add `RecurrentPPOTrainer`, or add alongside as a new file
- `src/env/games/mod.rs` — register new POMDP env

**To create (new files):**
- `src/policy/lstm.rs` — `LstmBurnPolicy<B>` and `LstmBurnConfig`
- `src/env/games/masked_cartpole.rs` (or `t_maze.rs`) — partially-observable environment
- `src/train/ppo/recurrent_trainer.rs` — `RecurrentPPOTrainer<B, P, O>` (if separate from existing trainer)
- `examples/recurrent_ppo_masked_cartpole.rs` — end-to-end training example

---

## Test Plan

- [ ] Unit test: `LstmBurnPolicy::forward_step` produces correct output shapes `[batch, hidden_dim]` for state tensors.
- [ ] Unit test: `LstmBurnPolicy` with same seed produces bit-identical weights across two constructions (mirrors `test_with_seed_is_bit_identical_orthogonal` in `mlp.rs`).
- [ ] Unit test: Hidden state resets to zeros when `terminated = true`; carries forward when `terminated = false`.
- [ ] Unit test: Sequence-preserving sampler produces contiguous subsequences with correct episode-boundary handling.
- [ ] Unit test: `RecurrentRolloutBuffer` round-trips hidden states correctly through `add_hidden_state` / retrieval.
- [ ] Integration test: `RecurrentPPOTrainer::train_step` runs without panic on a synthetic rank-3 observation batch.
- [ ] Integration test / example: Recurrent policy achieves learning on `MaskedCartPole` (or T-Maze) within N update iterations; feedforward baseline on same env fails to learn (optional but ideal for the issue's acceptance criterion).
- [ ] Regression: All existing feedforward PPO/A2C tests (`src/train/ppo/`, `src/train/a2c/`, `src/buffer/rollout/tests.rs`) pass without modification.

---

Curator note: The issue's own "Notes" section already signals that design-first is warranted. This analysis confirms it. Recommend converting to an epic (like #265/#266) with a design note as Phase 1 child issue before any implementation work begins. The buffer contract and minibatch strategy decisions are load-bearing — getting them wrong means reworking both the storage and trainer layers.
